pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -555,3 +555,48 @@ def is_module(self) -> bool:
         from pytorch_lightning import LightningDataModule, LightningModule  # prevent cyclic import
 
         return isinstance(self.instance, (LightningModule, LightningDataModule))
+
+
+@dataclass
+class _DataHookSelector:
+    """Stores the info about the shared DataHooks within LightningModule and LightningDataModule.
+
+    The hook source can be
+
+    1. a method from the :class:`~pytorch_lightning.core.lightning.LightningModule`,
+    2. a method from the :class:`~pytorch_lightning.core.datamodule.LightningDataModule`,
+
+    Arguments:
+        model: A LightningModule
+        datamodule: A LightningDataModule
+    """
+
+    model: "pl.LightningModule"
+    datamodule: Optional["pl.LightningDataModule"]
+    _valid_hooks: Tuple[str] = field(
+        default=("on_before_batch_transfer", "transfer_batch_to_device", "on_after_batch_transfer")
+    )
+
+    def get_hook(self, hook_name: str) -> Callable:
+        if hook_name not in self._valid_hooks:
+            raise ValueError(
+                f"`{hook_name}` is not a shared hook within `LightningModule` and `LightningDataModule`."
+                f" Valid hooks are {self._valid_hooks}."
+            )
+
+        if self.datamodule is None:
+            return getattr(self.model, hook_name)
+
+        if is_overridden(hook_name, self.datamodule):
+            if is_overridden(hook_name, self.model):
+                warning_cache.warn(
```

Comment:
I don't like this. I may not be the author of these implementations. I'm just reusing existing code from others and I can't really influence this. A warning sends me the signal that I could be doing something wrong.

Reply:
Not sure, I could be contradicting a previous reviewer here.

Reply:
but isn't that a good thing? If I am importing someone else's LM where all the training/eval logic is defined and not injecting my own LDM after overriding these hooks again, it should warn users that the old training logic rely on some additional logic as well defined under these hooks, now datamodule has overridden it so the old one will no longer be used at all.

Also, this is done for BC but now with a warning.
https://github.com/PyTorchLightning/pytorch-lightning/blob/1fa0639bcaf4a5ab058178b284c52ce5c13c1ca3/pytorch_lightning/trainer/connectors/data_connector.py#L204-L207

do you think we should raise an error now?